diff --git a/package/kernel/mt76/Makefile b/package/kernel/mt76/Makefile
index dd5c9f1..b85a3dc 100644
--- a/package/kernel/mt76/Makefile
+++ b/package/kernel/mt76/Makefile
@@ -6,7 +6,11 @@ PKG_RELEASE=4
 PKG_LICENSE:=GPLv2
 PKG_LICENSE_FILES:=
 
-PKG_VERSION:=wed
+PKG_SOURCE_URL:=https://github.com/openwrt/mt76
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_DATE:=2022-10-01
+PKG_SOURCE_VERSION:=72b87836d368685a093b927229209b61d8272541
+PKG_MIRROR_HASH:=ea4dfe93f7632d5063ac2ccf0f5a3698e2bfeadbbef45910381b5ef07e334cfc
 
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
 PKG_USE_NINJA:=0
