diff --git a/applications/luci-app-passwall/luasrc/model/cbi/passwall/api/gen_v2ray.lua b/applications/luci-app-passwall/luasrc/model/cbi/passwall/api/gen_v2ray.lua
index 9ad38ae..390b48d 100644
--- a/applications/luci-app-passwall/luasrc/model/cbi/passwall/api/gen_v2ray.lua
+++ b/applications/luci-app-passwall/luasrc/model/cbi/passwall/api/gen_v2ray.lua
@@ -293,13 +293,18 @@ if node_id then
     end
 
     if tcp_redir_port or udp_redir_port then
+        local sniffing
+        if remote_dns_fake then
+            sniffing = { enabled = true, destOverride = { "fakedns" }, metadataOnly = true }
+        else
+            sniffing = {enabled = sniffing and true or false, destOverride = { "http", "tls", (remote_dns_fake) and "fakedns"}, metadataOnly = false, routeOnly = route_only and true or nil, domainsExcluded = (sniffing and not route_only) and get_domain_excluded() or nil}
+        end
         local inbound = {
             protocol = "dokodemo-door",
             settings = {network = "tcp,udp", followRedirect = true},
             streamSettings = {sockopt = {tproxy = "tproxy"}},
-            sniffing = {enabled = sniffing and true or false, destOverride = {"http", "tls", (remote_dns_fake) and "fakedns"}, metadataOnly = false, routeOnly = route_only and true or nil, domainsExcluded = (sniffing and not route_only) and get_domain_excluded() or nil}
+            sniffing = sniffing,
         }
-    
         if tcp_redir_port then
             local tcp_inbound = api.clone(inbound)
             tcp_inbound.tag = "tcp_redir"
@@ -595,9 +600,9 @@ if remote_dns_server or remote_dns_doh_url or remote_dns_fake then
             tag = "dns-out",
             protocol = "dns",
             settings = {
-                address = remote_dns_server,
-                port = (_remote_dns_proto ~= "doh" and tonumber(remote_dns_port)) and tonumber(remote_dns_port) or 53,
-                network = "tcp",
+                address = "127.0.0.1",
+                port = 5356,
+                network = "udp",
             }
         })
 
diff --git a/applications/luci-app-passwall/root/etc/init.d/passwall b/applications/luci-app-passwall/root/etc/init.d/passwall
index 0767a50..3f98e3c 100755
--- a/applications/luci-app-passwall/root/etc/init.d/passwall
+++ b/applications/luci-app-passwall/root/etc/init.d/passwall
@@ -39,6 +39,7 @@ boot() {
 start() {
 	set_lock
 	[ $? == 1 ] && $APP_FILE echolog "脚本已经在运行，不重复运行，退出." && exit 0
+	/usr/sbin/ipset restore -file /usr/share/passwall/rules/custom_proxy_ips -!
 	$APP_FILE start
 	unset_lock
 }
diff --git a/applications/luci-app-passwall/root/usr/share/passwall/app.sh b/applications/luci-app-passwall/root/usr/share/passwall/app.sh
index 9cfdcd1..4b2c382 100755
--- a/applications/luci-app-passwall/root/usr/share/passwall/app.sh
+++ b/applications/luci-app-passwall/root/usr/share/passwall/app.sh
@@ -901,7 +901,7 @@ node_switch() {
 
 		#uci set $CONFIG.@global[0].${flag}_node=$new_node
 		#uci commit $CONFIG
-		source $APP_PATH/helper_${DNS_N}.sh logic_restart no_log=1
+		# source $APP_PATH/helper_${DNS_N}.sh logic_restart no_log=1
 	}
 }
 
@@ -1182,10 +1182,10 @@ start_dns() {
 		echolog "  + 过滤服务：ChinaDNS-NG(:${china_ng_listen_port})：国内DNS：${china_ng_chn}，可信DNS：${china_ng_gfw}"
 	}
 	
-	[ "$DNS_SHUNT" = "dnsmasq" ] && {
-		source $APP_PATH/helper_dnsmasq.sh stretch
-		source $APP_PATH/helper_dnsmasq.sh add FLAG="default" DNS_MODE=$DNS_MODE TMP_DNSMASQ_PATH=$TMP_DNSMASQ_PATH DNSMASQ_CONF_FILE=/tmp/dnsmasq.d/dnsmasq-passwall.conf REMOTE_FAKEDNS=$fakedns DEFAULT_DNS=$DEFAULT_DNS LOCAL_DNS=$LOCAL_DNS TUN_DNS=$TUN_DNS CHINADNS_DNS=$china_ng_listen TCP_NODE=$TCP_NODE PROXY_MODE=${TCP_PROXY_MODE}${LOCALHOST_TCP_PROXY_MODE}${ACL_TCP_PROXY_MODE} NO_PROXY_IPV6=${filter_proxy_ipv6}
-	}
+	# [ "$DNS_SHUNT" = "dnsmasq" ] && {
+	# 	source $APP_PATH/helper_dnsmasq.sh stretch
+	# 	source $APP_PATH/helper_dnsmasq.sh add FLAG="default" DNS_MODE=$DNS_MODE TMP_DNSMASQ_PATH=$TMP_DNSMASQ_PATH DNSMASQ_CONF_FILE=/tmp/dnsmasq.d/dnsmasq-passwall.conf REMOTE_FAKEDNS=$fakedns DEFAULT_DNS=$DEFAULT_DNS LOCAL_DNS=$LOCAL_DNS TUN_DNS=$TUN_DNS CHINADNS_DNS=$china_ng_listen TCP_NODE=$TCP_NODE PROXY_MODE=${TCP_PROXY_MODE}${LOCALHOST_TCP_PROXY_MODE}${ACL_TCP_PROXY_MODE} NO_PROXY_IPV6=${filter_proxy_ipv6}
+	# }
 }
 
 gen_pdnsd_config() {
@@ -1422,7 +1422,7 @@ start() {
 			start_redir UDP
 			start_dns
 			source $APP_PATH/iptables.sh start
-			source $APP_PATH/helper_${DNS_N}.sh logic_restart
+			# source $APP_PATH/helper_${DNS_N}.sh logic_restart
 		fi
 	}
 	start_crontab
@@ -1439,9 +1439,9 @@ stop() {
 	unset V2RAY_LOCATION_ASSET
 	unset XRAY_LOCATION_ASSET
 	stop_crontab
-	source $APP_PATH/helper_smartdns.sh del
-	source $APP_PATH/helper_dnsmasq.sh del
-	source $APP_PATH/helper_dnsmasq.sh restart no_log=1
+	# source $APP_PATH/helper_smartdns.sh del
+	# source $APP_PATH/helper_dnsmasq.sh del
+	# source $APP_PATH/helper_dnsmasq.sh restart no_log=1
 	rm -rf ${TMP_PATH}
 	rm -rf /tmp/lock/${CONFIG}_script.lock
 	echolog "清空并关闭相关程序和缓存完成。"
diff --git a/applications/luci-app-passwall/root/usr/share/passwall/iptables.sh b/applications/luci-app-passwall/root/usr/share/passwall/iptables.sh
index 01ade21..478f6d2 100755
--- a/applications/luci-app-passwall/root/usr/share/passwall/iptables.sh
+++ b/applications/luci-app-passwall/root/usr/share/passwall/iptables.sh
@@ -836,6 +836,39 @@ dns_hijack() {
 
 add_firewall_rule() {
 	echolog "开始加载防火墙规则..."
+
+	$ipt_m -N PSW_DIVERT
+	$ipt_m -A PSW_DIVERT -j MARK --set-mark 1
+	$ipt_m -A PSW_DIVERT -j ACCEPT
+
+	$ipt_m -N PSW_RULE
+	$ipt_m -A PSW_RULE -j CONNMARK --restore-mark
+	$ipt_m -A PSW_RULE -m mark --mark 1 -j RETURN
+	$ipt_m -A PSW_RULE -p tcp -m tcp --tcp-flags FIN,SYN,RST,ACK SYN -j MARK --set-xmark 1
+	$ipt_m -A PSW_RULE -p udp -m conntrack --ctstate NEW -j MARK --set-xmark 1
+	$ipt_m -A PSW_RULE -j CONNMARK --save-mark
+
+	$ipt_m -N PSW
+	insert_rule_before "$ipt_m" "PREROUTING" "mwan3" "-j PSW"
+	insert_rule_before "$ipt_m" "PREROUTING" "PSW" "-p tcp -m socket -j PSW_DIVERT"
+	$ipt_m -A PSW -p tcp $(dst proxy_ips) -j PSW_RULE
+	$ipt_m -A PSW -p udp $(dst proxy_ips) -j PSW_RULE
+	$ipt_m -A PSW -p tcp -m mark --mark 1 $(REDIRECT $TCP_REDIR_PORT TPROXY)
+	$ipt_m -A PSW -p udp -m mark --mark 1 $(REDIRECT $TCP_REDIR_PORT TPROXY)
+
+	$ipt_m -N PSW_OUTPUT
+	insert_rule_before "$ipt_m" "OUTPUT" "mwan3" "$(comment mangle-OUTPUT-PSW) -p tcp -j PSW_OUTPUT"
+	insert_rule_before "$ipt_m" "OUTPUT" "mwan3" "$(comment mangle-OUTPUT-PSW) -p udp -j PSW_OUTPUT"
+	$ipt_m -A PSW_OUTPUT -m mark --mark 0xff -j RETURN
+	$ipt_m -A PSW_OUTPUT -p tcp $(dst proxy_ips) -j PSW_RULE
+	$ipt_m -A PSW_OUTPUT -p udp $(dst proxy_ips) -j PSW_RULE
+
+	ip rule add fwmark 1 lookup 100
+	ip route add local 0.0.0.0/0 dev lo table 100
+}
+
+_add_firewall_rule() {
+	echolog "开始加载防火墙规则..."
 	ipset -! create $IPSET_LANIPLIST nethash maxelem 1048576
 	ipset -! create $IPSET_VPSIPLIST nethash maxelem 1048576
 	ipset -! create $IPSET_SHUNTLIST nethash maxelem 1048576
