diff --git a/applications/luci-app-zerotier/root/etc/init.d/zerotier b/applications/luci-app-zerotier/root/etc/init.d/zerotier
index 666d6753..72c28a1e 100755
--- a/applications/luci-app-zerotier/root/etc/init.d/zerotier
+++ b/applications/luci-app-zerotier/root/etc/init.d/zerotier
@@ -26,27 +26,40 @@ start_instance() {
 		echo "disabled in config"
 		return 1
 	fi
-  
-  [ -d /etc/config/zero ] || mkdir -p /etc/config/zero
-  config_path=/etc/config/zero
-  
-	config_get_bool port $cfg 'port'
+
+	config_get config_path $cfg 'config_path'
+	config_get port $cfg 'port'
 	config_get secret $cfg 'secret'
+	config_get local_conf $cfg 'local_conf'
+	config_get_bool copy_config_path $cfg 'copy_config_path' 0
+
+	path=${CONFIG_PATH}_$cfg
 
 	# Remove existing link or folder
-	rm -rf $CONFIG_PATH
+	rm -rf $path
 
-	# Create link from CONFIG_PATH to config_path
-	if [ -n "$config_path" -a "$config_path" != $CONFIG_PATH ]; then
+	# Create link or copy files from CONFIG_PATH to config_path
+	if [ -n "$config_path" -a "$config_path" != "$path" ]; then
 		if [ ! -d "$config_path" ]; then
-			echo "ZeroTier config_path does not exist: $config_path"
+			echo "ZeroTier config_path does not exist: $config_path" 1>&2
 			return
 		fi
 
-		ln -s $config_path $CONFIG_PATH
+		# ensure that the target exists
+		mkdir -p $(dirname $path)
+
+		if [ "$copy_config_path" = "1" ]; then
+			cp -r $config_path $path
+		else
+			ln -s $config_path $path
+		fi
 	fi
 
-	mkdir -p $CONFIG_PATH/networks.d
+	mkdir -p $path/networks.d
+
+	# link latest default config path to latest config path
+	rm -f $CONFIG_PATH
+	ln -s $path $CONFIG_PATH
 
 	if [ -n "$port" ]; then
 		ARGS="$ARGS -p$port"
@@ -56,7 +69,7 @@ start_instance() {
 		echo "Generate secret - please wait..."
 		local sf="/tmp/zt.$cfg.secret"
 
-		zerotier-idtool generate "$sf" > /dev/null
+		zerotier-idtool generate "$sf" >/dev/null
 		[ $? -ne 0 ] && return 1
 
 		secret="$(cat $sf)"
@@ -67,14 +80,18 @@ start_instance() {
 	fi
 
 	if [ -n "$secret" ]; then
-		echo "$secret" > $CONFIG_PATH/identity.secret
+		echo "$secret" >$path/identity.secret
 		# make sure there is not previous identity.public
-		rm -f $CONFIG_PATH/identity.public
+		rm -f $path/identity.public
+	fi
+
+	if [ -f "$local_conf" ]; then
+		ln -s "$local_conf" $path/local.conf
 	fi
 
 	add_join() {
 		# an (empty) config file will cause ZT to join a network
-		touch $CONFIG_PATH/networks.d/$1.conf
+		touch $path/networks.d/$1.conf
 	}
 
 	config_list_foreach $cfg 'join' add_join
@@ -88,18 +105,17 @@ start_instance() {
 start_service() {
 	config_load 'zerotier'
 	config_foreach start_instance 'zerotier'
-	touch /tmp/zero.log && /etc/zerotier.start > /tmp/zero.log 2>&1 &
+	touch /tmp/zero.log && /etc/zerotier.start >/tmp/zero.log 2>&1 &
 }
 
 stop_instance() {
-  rm -f /tmp/zero.log
+	rm -f /tmp/zero.log
 	local cfg="$1"
 
-	/etc/zerotier.stop > /tmp/zero.log 2>&1 &
+	/etc/zerotier.stop >/tmp/zero.log 2>&1 &
 
 	# Remove existing link or folder
-	rm -f $CONFIG_PATH/networks.d/*.conf
-	rm -rf $CONFIG_PATH
+	rm -rf ${CONFIG_PATH}_${cfg}
 }
 
 stop_service() {
diff --git a/applications/luci-app-zerotier/root/etc/zerotier.start b/applications/luci-app-zerotier/root/etc/zerotier.start
index b43e5f97..46686f56 100755
--- a/applications/luci-app-zerotier/root/etc/zerotier.start
+++ b/applications/luci-app-zerotier/root/etc/zerotier.start
@@ -11,6 +11,12 @@ zero_enable="$(uci get zerotier.sample_config.enabled)"
 	done
 }
 
+iptables_checked_insert() {
+	table=$1
+	shift
+	iptables -t $table -C $@ || iptables -t $table -I $@
+}
+
 nat_enable="$(uci get zerotier.sample_config.nat)"
 zt0="$(ifconfig | grep 'zt' | awk '{print $1}')"
 echo "${zt0}" > "/tmp/zt.nif"
@@ -19,10 +25,10 @@ echo "${zt0}" > "/tmp/zt.nif"
 	for i in ${zt0}
 	do
 		ip_segment=""
-		iptables -I FORWARD -i "$i" -j ACCEPT
-		iptables -I FORWARD -o "$i" -j ACCEPT
-		iptables -t nat -I POSTROUTING -o "$i" -j MASQUERADE
+		iptables_checked_insert filter FORWARD -i "$i" -j ACCEPT
+		iptables_checked_insert filter FORWARD -o "$i" -j ACCEPT
+		iptables_checked_insert nat POSTROUTING -o "$i" -j MASQUERADE
 		ip_segment="$(ip route | grep "dev $i proto kernel" | awk '{print $1}')"
-		iptables -t nat -I POSTROUTING -s "${ip_segment}" -j MASQUERADE
+		iptables_checked_insert nat POSTROUTING -s "${ip_segment}" -j MASQUERADE
 	done
 }
diff --git a/applications/luci-app-zerotier/root/etc/zerotier/zerotier.log b/applications/luci-app-zerotier/root/etc/zerotier/zerotier.log
deleted file mode 100644
index e69de29b..00000000
